### PR TITLE
Connectable observable

### LINKED ISSUE #166: Add publish, replay, share, etc
In other Rx implementations, there are various operators to transform a stream into a multicast stream (`share`, `publish`) that automatically subscribes and unsubscribes the source stream (`refCount`) and repeats a certain number of items on subscription (`publishReplay`, `shareReplay`). Is there a way to do this with rxdart or the dart stream api?

The only thing I found is `asBroadcastStream()`. It doesn't replay values, and I'm not sure what happens when all subscribers unsubscribe.

https://www.learnrxjs.io/operators/multicasting/sharereplay.html

https://medium.com/mindorks/how-to-use-rxjava-share-operator-26b08973771a

### PR DESCRIPTION
**Pull request by brianegan**
Introduces the concept of a `ConnectableObservable` and operators for converting normal Streams into `ConnectableObservables` or Broadcast Streams that are backed up by Subjects.

It also introduces the concept of `BehaviorObservable`, which is an observable that provides sync access to the latest value. This is used for `BehaviorSubject`. Furthermore, introduces the idea of a `ReplayObservable,` which provides sync access to the emitted values from a `ReplaySubject`.

Added operators:

  - publish
  - publishBehavior
  - publishReplay
  - share
  - shareBehavior
  - shareReplay

This is based on the ConnectableObservable class from RxJava and RxJS. Operator names inspired by those versions of Rx, but tailored to the Dart environment.

Overall, these tools allow you to convert a single-subscription Observable into a broadcast observable, backed up by the appropriate Subject. This prevents single-subscription errors, and gives AngularDart / Flutter apps the ability to read the latest value from a converted Observable synchronously. This is useful for `async` pipes and `StreamBuilder` widgets.

Fixes #166

### CODE REVIEW

**Comment by sir-boformer** [comment]
This looks very solid. Clean and simple building blocks.

File: test/observables/behavior_connectable_observable_test.dart [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    test('single subscription error', () async {
+      final ConnectableObservable<int> observable =
+          BehaviorConnectableObservable<int>(
+              Stream<int>.fromIterable(<int>[1, 2, 3]));
+
+      observable.connect();
+      observable.listen(null)..cancel();
+
+      expect(observable, emitsInOrder(<int>[1, 2, 3]));
```
[the remark this answers is not in the document]
Comment (brianegan) [reply test/observables/behavior_connectable_observable_test.dart]: Whoops, meant to delete this test...

File: test/observables/behavior_connectable_observable_test.dart [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    });
+
+    test('keeps emitting with an active subscription', () async {
+      final Observable<int> observable =
+          Observable<int>.fromIterable(<int>[1, 2, 3]).shareBehavior();
+
+      observable.listen(null);
+      observable.listen(null)..cancel();
+
+      expect(observable, emitsInOrder(<int>[1, 2, 3]));
```
[the remark this answers is not in the document]
Comment (brianegan) [reply test/observables/behavior_connectable_observable_test.dart]: This is an interesting one -- all of the subscriptions will be setup synchronously before the `Stream.fromIterable` starts emitting items asynchronously.

I've got another test that verifies a "Late" subscription will only get `3`.

**Comment by brianegan** [comment]
Thanks for the review @sir-boformer! Comments addressed

**Comment by sir-boformer** [comment]
I'm going to test this in development and will report back all issues.

**Comment by long1eu** [comment]
This works awesome, but it doesn't reconnect after the subscribers drop to 0 and then they listen again. Is this as intended?

**Comment by brianegan** [comment]
@long1eu It is -- Unfortunately, it is. It would be great if we could support re-connecting, but unlike plain old Rx Observables, Many Streams are Single-subscription only. Therefore, we can't automatically reconnect or in many cases we'd get an Error like "Stream has already been listened to."

**Comment by brianegan** [comment]
Yah, I was considering that as well, but pausing can might cause unexpected results... 

The `StreamSubscription` will buffer all events while the subscription is paused, and replay them once the connection is resumed. I'm not sure the consumers will expect a replay of all buffered events. We could debounce the events internally, but I'm not 100% sure that's a great idea

**Comment by brianegan** [comment]
Hey all -- quick heads up: I've renamed `BehaviorObservable` back to `ValueObservable` based on @filiph's feedback. Hopefully that will be easier to understand for folks, especially peeps coming from a Flutter background where `ValueNotifier` is pretty common.

**Pull request merged**